src/darwin/Framework/CHIP/MTRAttributeValueWaiter.mm

```diff
@@ -120,15 +143,41 @@ - (void)_notifyWithError:(NSError * _Nullable)error
         MTR_LOG("%@ %p wait for attribute values unknown error: %@", self, self, error);
     }
 
-    auto completion = self.completion;
-    auto queue = self.queue;
-    self.completion = nil;
-    self.queue = nil;
     dispatch_async(queue, ^{
         completion(error);
     });
 }
 
+- (void)_startTimerWithTimeout:(NSTimeInterval)timeout
+{
+    // Have the timer dispatch on the device queue, so we are not trying to do
+    // it on the API consumer queue (which might be blocked by the API
+    // consumer).
+    dispatch_source_t timerSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.device.queue);
+
+    // Set a timer to go off after timeout, and not repeat.
+    dispatch_source_set_timer(timerSource, dispatch_time(DISPATCH_TIME_NOW, static_cast<uint64_t>(timeout * static_cast<double>(NSEC_PER_SEC))), DISPATCH_TIME_FOREVER,
+        // Allow .5 seconds of leeway; should be plenty, in practice.
+        static_cast<uint64_t>(0.5 * static_cast<double>(NSEC_PER_SEC)));
+
+    mtr_weakify(self);
+    dispatch_source_set_event_handler(timerSource, ^{
+        dispatch_source_cancel(timerSource);
+        mtr_strongify(self);
+        if (self != nil) {
+            [self.device _forgetAttributeWaiter:self];
+            [self _notifyWithError:[MTRError errorForCHIPErrorCode:CHIP_ERROR_TIMEOUT]];
+        }
+    });
+
+    {
```

Comment:
Locking here isn't even strictly necessary right? The timer can't kick in until it's resumed, and the waiter hasn't been returned to the client yet.

Reply:
True, but that global state is non-obvious from here.  So it's safer to just do our invariants, which is locking around the relevant field.